aidial_adapter_bedrock/llm/chat_model.py

```diff
@@ -126,24 +122,33 @@ async def chat(
 
     async def truncate_prompt(
         self, params: ModelParameters, messages: List[Message]
-    ) -> List[int]:
-        prompt = self.get_text_completion_prompt(params, messages)
-        return prompt.discarded_messages or []
+    ) -> List[int] | None:
+        prompt = await self.get_text_completion_prompt(params, messages)
+        return prompt.discarded_messages
 
 
-def default_keep_message(messages: List[BaseMessage], idx: int) -> bool:
-    """Keep system messages and the last message."""
-    return isinstance(messages[idx], SystemMessage) or idx == len(messages) - 1
+def keep_last(messages: List[Any], idx: int) -> bool:
+    return idx == len(messages) - 1
 
 
-def default_partitioner(messages: List[BaseMessage]) -> List[int]:
+def keep_last_and_system_messages(
+    messages: List[BaseMessage], idx: int
+) -> bool:
+    return isinstance(messages[idx], SystemMessage) or keep_last(messages, idx)
+
+
+def trivial_partitioner(messages: List[Any]) -> List[int]:
     return [1] * len(messages)
 
 
+def turn_based_partitioner(messages: List[Any]) -> List[int]:
```

Comment:
Could you please elaborate, why you came to this idea of turn based partition during truncation?

Currently, I see that this approach can only remove such case

```
# messages from last to first

model_answer_10
user_prompt_10
model_answer_9
user_prompt_9
model_answer_8
<---- Here is our limit ---- >
user_answer_8
```

With simple approach without partitions we'll include `model_answer_8` without `user_answer_8`, with partitions we'll have both `model_answer_8` and `user_answer_8` trimmed

But is it such beneficial to introduce this complicated `partitions` logic?

Tbh I don't even know if that's beneficial - if we don't user partitions, we provide more information to model, and more tokens to the sequence

Reply:
> But is it such beneficial to introduce this complicated partitions logic?

It's not beneficial, it's required, because Gemini model could only accept a conversation with a valid sequence of User-Bot turns:

```
System-[User-Bot]-[User-Bot]-User: is ok,
System-[Bot!]-[User-Bot]-User: fails Gemini request validation
```